File: server/src/main/java/gg/mineral/server/network/packet/login/serverbound/EncryptionKeyResponsePacket.java
package gg.mineral.server.network.packet.login.serverbound;

import gg.mineral.api.network.connection.Connection;
import gg.mineral.api.network.packet.Packet;
import gg.mineral.server.network.connection.ConnectionImpl;
import io.netty.buffer.ByteBuf;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.val;
import lombok.experimental.Accessors;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Accessors(fluent = true)
public final class EncryptionKeyResponsePacket implements Packet.INCOMING {
    private byte[] sharedSecretBytes, verifyToken;

    @Override
    public void received(Connection connection) {
        if (connection instanceof ConnectionImpl impl) {
            boolean success = impl.authenticate(sharedSecretBytes, verifyToken);
            val config = impl.getServer().getConfig();

            if (success)
                try {
                    impl.loggedIn();
                } catch (Exception e) {
                    e.printStackTrace();
                    connection.disconnect(config.getDisconnectCanNotAuthenticate());
                }
            else
                connection.disconnect(config.getDisconnectCanNotAuthenticate());
        }
    }

    @Override
    public void deserialize(ByteBuf is) {
        short lengthOfSharedSecret = is.readShort();
        sharedSecretBytes = new byte[lengthOfSharedSecret];
        is.readBytes(sharedSecretBytes);
        short lengthOfVerifyToken = is.readShort();
        verifyToken = new byte[lengthOfVerifyToken];
        is.readBytes(verifyToken);
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/login/serverbound/EncryptionKeyResponsePacket.kt
package gg.mineral.server.network.packet.login.serverbound

import gg.mineral.api.network.connection.Connection
import gg.mineral.api.network.packet.Packet
import gg.mineral.server.network.connection.ConnectionImpl
import gg.mineral.server.network.protocol.ProtocolState
import io.netty.buffer.ByteBuf
import kotlinx.coroutines.withContext

class EncryptionKeyResponsePacket(
    var sharedSecretBytes: ByteArray? = null,
    var verifyToken: ByteArray? = null
) : Packet.Incoming, Packet.AsyncHandler {
    override suspend fun receivedAsync(connection: Connection) {
        if (connection !is ConnectionImpl) return
        val secretKey = connection.authenticate(sharedSecretBytes!!, verifyToken!!)

        val config = connection.server.config
        if (secretKey != null) {
            try {
                connection.enableEncryption(secretKey)
                connection.loginSuccess()
            } catch (e: Exception) {
                e.printStackTrace()
                connection.disconnect(config.disconnectCanNotAuthenticate)
            }
        } else connection.disconnect(config.disconnectCanNotAuthenticate)

        withContext(connection.server.syncDispatcher) {
            if (connection.protocolState == ProtocolState.PLAY) connection.spawnPlayer()
        }
    }

    override fun deserialize(`is`: ByteBuf) {
        val lengthOfSharedSecret = `is`.readShort()
        sharedSecretBytes = ByteArray(lengthOfSharedSecret.toInt())
        `is`.readBytes(sharedSecretBytes)
        val lengthOfVerifyToken = `is`.readShort()
        verifyToken = ByteArray(lengthOfVerifyToken.toInt())
        `is`.readBytes(verifyToken)
    }
}
